includes/data-stores/class-wc-admin-reports-data-store.php

```diff
@@ -559,26 +567,151 @@ protected function get_products_by_cat_ids( $categories ) {
+	 * Returns comma separated ids of allowed products, based on query arguments from the user.
 	 *
 	 * @param array $query_args Parameters supplied by the user.
-	 * @return array
+	 * @return string
 	 */
-	protected function get_allowed_products( $query_args ) {
-		$allowed_products = array();
+	protected function get_included_products( $query_args ) {
+		$included_products = array();
+		$operator          = $this->get_match_operator( $query_args );
+
 		if ( isset( $query_args['categories'] ) && is_array( $query_args['categories'] ) && count( $query_args['categories'] ) > 0 ) {
-			$allowed_products = $this->get_products_by_cat_ids( $query_args['categories'] );
-			$allowed_products = wc_list_pluck( $allowed_products, 'get_id' );
+			$included_products = $this->get_products_by_cat_ids( $query_args['categories'] );
+			$included_products = wc_list_pluck( $included_products, 'get_id' );
 		}
 
-		if ( isset( $query_args['products'] ) && is_array( $query_args['products'] ) && count( $query_args['products'] ) > 0 ) {
-			if ( count( $allowed_products ) > 0 ) {
-				$allowed_products = array_intersect( $allowed_products, $query_args['products'] );
+		if ( isset( $query_args['product_includes'] ) && is_array( $query_args['product_includes'] ) && count( $query_args['product_includes'] ) > 0 ) {
+			if ( count( $included_products ) > 0 ) {
+				if ( 'AND' === $operator ) {
+					$included_products = array_intersect( $included_products, $query_args['product_includes'] );
+				} elseif ( 'OR' === $operator ) {
+					// Union of products from selected categories and manually included products.
+					$included_products = array_unique( array_merge( $included_products, $query_args['product_includes'] ) );
+				}
 			} else {
-				$allowed_products = $query_args['products'];
+				$included_products = $query_args['product_includes'];
+			}
+		}
+
+		$included_products_str = implode( ',', $included_products );
+		return $included_products_str;
+	}
+
+	/**
+	 * Returns comma separated ids of excluded products, based on query arguments from the user.
+	 *
+	 * @param array $query_args Parameters supplied by the user.
+	 * @return string
+	 */
+	protected function get_excluded_products( $query_args ) {
+		$excluded_products_str = '';
+
+		if ( isset( $query_args['product_excludes'] ) && is_array( $query_args['product_excludes'] ) && count( $query_args['product_excludes'] ) > 0 ) {
+			$excluded_products_str = implode( ',', $query_args['product_excludes'] );
+		}
+		return $excluded_products_str;
+	}
+
+	/**
+	 * Returns comma separated ids of included coupons, based on query arguments from the user.
+	 *
+	 * @param array $query_args Parameters supplied by the user.
+	 * @return string
+	 */
+	protected function get_included_coupons( $query_args ) {
+		$included_coupons_str = '';
+
+		if ( isset( $query_args['coupon_includes'] ) && is_array( $query_args['coupon_includes'] ) && count( $query_args['coupon_includes'] ) > 0 ) {
```

Comment:
Coupons filter isn't working. Maybe because `coupon_includes` is used in this function, but `code_includes` is specified in the schema and `prepare_reports_query`.

When we spoke about this last time, we decided `coupon_includes` is best, if I recall correctly.

Reply:
Yup, fixed, thanks.